File: inJava/app/src/main/java/com/barrygithub/rickandmortyapp/data/RemoteDatasource/entities/EntityApi.java
package com.barrygithub.rickandmortyapp.data.RemoteDatasource.entities;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;

import com.google.gson.annotations.SerializedName;

import java.util.List;

/**
 * Project RickAndMortyApp
 * Created by Barry Zea H. on 3/04/23.
 * Copyright (c) Barry Zea H. All rights reserved.
 **/
public class EntityApi implements Parcelable {
    private @SerializedName("info") MetaData metaData;
    private @SerializedName("results") List<CharacterApi> results;

    public EntityApi() {
    }

    protected EntityApi(Parcel in) {
        metaData = in.readParcelable(MetaData.class.getClassLoader());
        results = in.createTypedArrayList(CharacterApi.CREATOR);
    }

    public static final Creator<EntityApi> CREATOR = new Creator<EntityApi>() {
        @Override
        public EntityApi createFromParcel(Parcel in) {
            return new EntityApi(in);
        }

        @Override
        public EntityApi[] newArray(int size) {
            return new EntityApi[size];
        }
    };

    public MetaData getMetaData() {
        return metaData;
    }

    public void setMetaData(MetaData metaData) {
        this.metaData = metaData;
    }

    public List<CharacterApi> getResults() {
        return results;
    }

    public void setResults(List<CharacterApi> results) {
        this.results = results;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(@NonNull Parcel parcel, int i) {
        parcel.writeParcelable(metaData, i);
        parcel.writeTypedList(results);
    }


}

File: inKotlin/app/src/main/java/com/barrygithub/rickandmortyapp/data/remoteDatasource/entities/EntityApi.kt
package com.barrygithub.rickandmortyapp.data.remoteDatasource.entities

import com.barrygithub.rickandmortyapp.data.localDatasource.Character
import com.barrygithub.rickandmortyapp.data.localDatasource.EntityDb
import com.google.gson.annotations.SerializedName

/**
 * Project RickAndMortyApp
 * Created by Barry Zea H. on 14/04/23.
 * Copyright (c) Barry Zea H. All rights reserved.
 *
 **/

data class EntityApi(
    @SerializedName("info") var metaData:MetaData = MetaData(),
    @SerializedName("results") var results:List<CharacterApi> = arrayListOf()
)

data class MetaData(
    @SerializedName("count") var count:Int=0,
    @SerializedName("pages") var pages:Int=0
)

data class CharacterApi(
    @SerializedName("id") var id:Int=0,
    @SerializedName("name") var name:String,
    @SerializedName("status") var status:String="",
    @SerializedName("species")var species:String="",
    @SerializedName("gender") var gender:String="",
    @SerializedName("type") var type:String="",
    @SerializedName("image") var image:String="",
    @SerializedName("episode") var episode:List<String> = arrayListOf()
)
data class Episode(
    @SerializedName("name") var name:String="",
    @SerializedName("air_date")var airDate:String="",
    @SerializedName("episode") var episode:String=""
)

fun EntityApi.toDatabase() = EntityDb(
    info = com.barrygithub.rickandmortyapp.data.localDatasource.MetaData(
        count = metaData.count,
        pages = metaData.pages
    ),
    results = results.map { Character(
        id=it.id,
        name=it.name,
        status=it.status,
        species = it.species, gender = it.gender,
        type = it.type,
        image = it.image, episodes = it.episode) }

)


